add a metalad description (#930)
datalad-handbook/book: docs/beyond_basics/101-181-metalad.rst

--------------------------------

For many years, :term:`metadata` related functionality was included in the DataLad core package.

> A modernized approach, however, is now developed in the `datalad-metalad extension <http://docs.datalad.org/projects/metalad/en/latest/>`_.

Reviewer: Should we rather point to the github repository here? At least while the documentation is still outdated.

```suggestion
A modernized approach, however, is now developed in the `datalad-metalad extension <https://github.com/datalad/datalad-metalad>`_.
```

Reviewer 2: yeah, I guess its better than confusing people :shrug: thx for the suggestion

To allow a variety of metadata use cases, MetaLad can use various metadata schemas simultaneously - if you want to, all schemas from the example above and many more could be created and managed in the same dataset in parallel.

To handle different schemas in parallel, MetaLad represents them based on unique identifiers of the extraction process that generated them.
The automatically scanned full text might be identified with an extractor name ``OCR``, and the citation data could be called "`altmetric <https://en.wikipedia.org/wiki/Altmetric>`_".
But while the term "extractor" has a technical feel to it, an "extractor" can also be the manual process of annotating arbitrary information about a file - nothing prevents metadata from medical annotations to be called ``Sam-tracing-brain-regions-by-hand``.

In addition to identifying schemas via extractor names, MetaLad and other :term:`DataLad extension`\s ship with specialized extractor tools to extract metadata of a certain schema.

> Likewise, anyone can build their own extractor to generate schemas of their choice.

Reviewer 2: yes!

> Likewise, anyone can build their own extractor to generate schemas of their choice.

Reviewer 2: I'm using docs.datalad.org/projects/metalad/user_guide/writing-extractors.html as the prospective URL, right?

> Likewise, anyone can build their own extractor to generate schemas of their choice.

Reviewer: yes, that makes sense to me

> Likewise, anyone can build their own extractor to generate schemas of their choice.

Reviewer 2: thx :)

The following call would add the metadata entry to the current dataset, ``cozy-screensavers``::

    $ datalad meta-add -d cozy-screensavers - < metadata-zen.json

.. find-out-more:: meta-add validity checks

	When adding metadata for the first time, its not uncommon to run into errors.
	Its quite easy, for example, to miss a comma or quotation mark when creating a JSON object by hand.
	But there are also some internal checks that might be surprising.
	If you want to add the metadata above to your own dataset, you should make sure to adjust the ``dataset_id`` to the ID of your own dataset, found via the command ``datalad configuration get datalad.dataset.id`` - otherwise you'll see an error [#f4]_.
	And in case you'd supply the ``extraction_time`` as "this morning at 8AM" instead of a time stamp, the command will be unhappy as well.
	In case an error occurs, make sure to read the error message, and turn the the commands' ``--help`` for insights about requirements you might have missed.

After the metadata has been added, you can view it via the command :command:`meta-dump`.

> The simplest form of this command is ``meta-dump -r``, which will show all metadata that is stored in dataset in the current directory.

Reviewer: ```suggestion
The simplest form of this command is ``meta-dump -r``, which will show all metadata that is stored in the dataset in the current directory.
```

TODO: This needs and example

Creating your own extractor
"""""""""""""""""""""""""""

TODO: This also needs an example

Distributing and Getting Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Once metadata has been added to a dataset, it can be distributed and retrieved.
Instead of creating and adding metadata yourself, you could download fitting pre-existing metadata.
Similarly, instead of repeating a ``meta-add`` process for one and the same files across hierarchies of datasets, metadata added into one dataset can be exported into other datasets.
Regardless of whether it is a distribution or a retrieval process, though, an export with MetaLad will only concern the *metadata*, and never the primary data.

Download Metadata from a remote repository
""""""""""""""""""""""""""""""""""""""""""

Let's start by creating a place where someone else's metadata could live.

.. code-block:: bash

    $ datalad create metadata-assimilation
    $ cd metadata-assimilation

Because MetaLad stores metadata in :term:`Git`'s object store, we use Git to directly fetch metadata from a remote repository, such as this demo on :term:`GitHub`: ``https://github.com/christian-monch/metadata-test.git``.

> Because metadata added by MetaLad is not transported automatically but needs to be specifically requested, the command to retrieve it looks unfamiliar to non-Git-users: It identifies the precise location of :term:`ref` that contains the metadata.

Reviewer: ```suggestion
Because metadata added by MetaLad is not transported automatically but needs to be specifically requested, the command to retrieve it looks unfamiliar to non-Git-users: It identifies the precise location of the :term:`ref` that contains the metadata.
```

   * It should be possible to only retrieve parts of the overall metadata tree, e.g. certain sub-nodes

   TODO this needs much more

.. code-block:: bash

    $ git fetch \
      "https://github.com/christian-monch/metadata-test.git" \
      "refs/datalad/*:refs/datalad/*"

The metadata is now locally available in the Git repository ``metadata-repo``.
You can verify this by issuing the command ``datalad meta-dump -r``, which will list all metadata in the repository.

Publish metadata to a Git-Repository
""""""""""""""""""""""""""""""""""""

You can also push your metadata to a remote :term:`sibling` (if you have write :term:`permissions`).
This, too, uses a Git command to push only specific :term:`ref`\s.
Let's assume you are in the directory that contains the git repository with your metadata, then you can push your metadata to a remote git repository ``<your repository>``::

    $ git push "<your repository>" "refs/datalad/*:refs/datalad/*"

You will notice that no primary data is stored in the repository ``metadata-destination``. That allows you to publish metadata without publishing the primary data at the same time.

> Querying metadata

Reviewer: Perhaps this would be a good place to add some note about and links to the catalog? Or a similar section titled something like `Using metadata`

Reviewer: @adswa E.g. under a `Using metadata` section (or `Let's put it to use`) we could say something along the lines of:

- now that we know all about metadata and how it is handled by metalad, a final note on its utility
- metadata, especially when originating from different sources and harmonized to a single schema, provide the powerful opportunity to aid data discoverability
- an example of a good use case for metadata is therefore search or browsing interfaces or data bases, such as data portals graph query databases, etc
- Metalad-extracted metadata can be used in workflows to generate such interfaces, an example being datalad catalog
- the example: add metadata to datalad datasets, extract metadata with metalad and multiple extractors, translate extracted metadata to the catalog schema, submit it to datalad-catalog in order to generate catalog entries, which can all be browsed in a user friendly web-based interface.

Reviewer 2: I have added this pretty much exactly in your words - thanks for the excellent suggestion!

	In case an error occurs, make sure to read the error message, and turn the the commands' ``--help`` for insights about requirements you might have missed.

After the metadata has been added, you can view it via the command :command:`meta-dump`.
The simplest form of this command is ``meta-dump -r``, which will show all metadata that is stored in the dataset in the current directory.
To get more specific metadata records, you can give a dataset-file-path-pattern to ``meta-dump``, much like an argument to ``ls``, that identifies :term:`dataset ID`, version and a file within the dataset.
The two parts are separated by ``:``. The following line would just dump all metadata for ``zen.png``.

.. runrecord:: _examples/DL-101-181-106
   :language: console
   :workdir: beyond_basics/meta/cozy-screensavers

   $ datalad meta-dump -d . .:zen.png

This could also be printed a bit more readable:

.. runrecord:: _examples/DL-101-181-107
   :language: console
   :workdir: beyond_basics/meta/cozy-screensavers

   $ datalad -f json_pp meta-dump -d . .:zen.png

.. find-out-more:: More complex metadata-dumps

   TODO: add complex Dataset-file-path-pattern examples, e.g., with UUIDs, versions, etc

> 

Reviewer 2: Does anyone have complex dumps from the real-world? The docstring of meta-add contains the spec for this notation, but it has a complex-ish feel to it, so the more examples, the better :)

Using existing extractors to add metadata
"""""""""""""""""""""""""""""""""""""""""

> TODO: This needs and example

Reviewer 2: @christian-monch do you by chance have, e.g., the call how you used the study-mini-meta extractor? or shall I just cure the core extractor included in metalad here?

Reviewer: Is this useful? I'm running it on a local version of the studyforrest superdataset which has a studyminimeta file added (see PR here: https://github.com/psychoinformatics-de/studyforrest-data/pull/61/files)

```bash
$ cd studyforrest-data
$ datalad meta-extract -d . metalad_studyminimeta > metadata_record.json

[INFO] Start studyminimeta metadata extraction from /Users/jsheunis/Documents/psyinf/Data/studyforrest-data
[INFO] Finished studyminimeta metadata extraction from /Users/jsheunis/Documents/psyinf/Data/studyforrest-data

$ cat extracted_metadata.json | jq .

{
  "type": "dataset",
  "dataset_id": "deabeb9b-7a37-4062-a1e0-8fcef7909609",
  "dataset_version": "6d7fb68264f9b9951ae141fc830712a8744e3293",
  "extractor_name": "metalad_studyminimeta",
  "extractor_version": "0.1",
  "extraction_parameter": {},
  "extraction_time": 1676280974.146493,
  "agent_name": "Stephan Heunis",
  "agent_email": "s.heunis@fz-juelich.de",
  "extracted_metadata": {
    "@context": {
      "@vocab": "http://schema.org/"
    },
    "@graph": [
      {
        "@id": "https://schema.datalad.org/datalad_dataset#deabeb9b-7a37-4062-a1e0-8fcef7909609",
        "@type": "Dataset",
        "version": "6d7fb68264f9b9951ae141fc830712a8744e3293",
        "name": "StudyForrest",
        "url": "https://github.com/psychoinformatics-de/studyforrest-data",
        "description": "The StudyForrest project centers around the use of the movie Forrest Gump, which provides complex sensory input that is both reproducible and is also richly laden with real-life-like content and contexts. Since its initial release, the StudyForrest dataset has grown and been extended substantially, and now encompasses many hours of fMRI scans, structural brain scans, eye-tracking data, and extensive annotations of the movie. It is a one-of-a-kind resource for studying high-level cognition in the human brain under complex, natural stimulation. The versatility of the provided data (some individuals have nearly ten hours of fMRI data) enables studies far beyond this main focus. This covers a vast range from studies of low-level signal properties and brain structure, to sensory integration and attentional processes, to computational modeling of representational spaces and brain area interactions.",
        "keywords": [
          "human",
          "fMRI",
          "task",
          "7T",
          "3T",
          "audio",
          "visual",
          "music",
          "retinotopy",
          "angiography",
          "T1,T2"
        ],
        "author": [
          {
            "@id": "https://schema.datalad.org/person#g@fz-juelich.de"
          }
        ],
        "funder": [
          {
            "@id": "https://schema.datalad.org/organization#BMBF, 01GQ1411",
            "@type": "Organization",
            "name": "BMBF, 01GQ1411"
          },
          {
            "@id": "https://schema.datalad.org/organization#NSF, 1429999",
            "@type": "Organization",
            "name": "NSF, 1429999"
          }
        ]
      },
      {
        "@id": "#publicationList",
        "@list": [
          {
            "@id": "#publication[0]",
            "@type": "ScholarlyArticle",
            "headline": "The effect of acquisition resolution on orientation decoding from V1: comparison of 3T and 7T",
            "datePublished": 2018,
            "sameAs": "https://doi.org/10.1101/305417",
            "author": [
              {
                "@id": "https://schema.datalad.org/person#a@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#b@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#c@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#d@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#e@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#f@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#g@fz-juelich.de"
              }
            ],
            "publication": {
              "@id": "https://schema.datalad.org/publication_event#bioRxiv",
              "@type": "PublicationEvent",
              "name": "bioRxiv"
            }
          },
          {
            "@id": "#publication[1]",
            "@type": "ScholarlyArticle",
            "headline": "Spatial band-pass filtering aids decoding musical genres from auditory cortex 7T fMRI",
            "datePublished": 2018,
            "sameAs": "https://doi.org/10.12688/f1000research.13689.2",
            "author": [
              {
                "@id": "https://schema.datalad.org/person#a@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#f@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#g@fz-juelich.de"
              }
            ],
            "publication": {
              "@id": "https://schema.datalad.org/publication_event#F1000Research",
              "@type": "PublicationEvent",
              "name": "F1000Research"
            }
          },
          {
            "@id": "#publication[2]",
            "@type": "ScholarlyArticle",
            "headline": "The effect of acquisition resolution on orientation decoding from V1 BOLD fMRI at 7 Tesla",
            "datePublished": 2017,
            "sameAs": "https://doi.org/10.1016/j.neuroimage.2016.12.040",
            "author": [
              {
                "@id": "https://schema.datalad.org/person#a@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#c@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#b@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#f@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#g@fz-juelich.de"
              }
            ],
            "publication": {
              "@id": "https://schema.datalad.org/publication_event#NeuroImage",
              "@type": "PublicationEvent",
              "name": "NeuroImage"
            }
          },
          {
            "@id": "#publication[3]",
            "@type": "ScholarlyArticle",
            "headline": "Ultra high-field multi-resolution fMRI data for orientation decoding in visual cortex",
            "datePublished": 2017,
            "sameAs": "https://doi.org/10.1016/j.dib.2017.05.014",
            "author": [
              {
                "@id": "https://schema.datalad.org/person#a@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#b@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#c@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#d@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#e@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#f@fz-juelich.de"
              },
              {
                "@id": "https://schema.datalad.org/person#g@fz-juelich.de"
              }
            ],
            "publication": {
              "@id": "https://schema.datalad.org/publication_event#Data in Brief",
              "@type": "PublicationEvent",
              "name": "Data in Brief"
            }
          }
        ]
      },
      {
        "@id": "#personList",
        "@list": [
          {
            "@id": "https://schema.datalad.org/person#a@fz-juelich.de",
            "@type": "Person",
            "email": "a@fz-juelich.de",
            "name": " Ayan Sengupta",
            "givenName": "Ayan",
            "familyName": "Sengupta"
          },
          {
            "@id": "https://schema.datalad.org/person#b@fz-juelich.de",
            "@type": "Person",
            "email": "b@fz-juelich.de",
            "name": " Oliver Speck",
            "givenName": "Oliver",
            "familyName": "Speck"
          },
          {
            "@id": "https://schema.datalad.org/person#c@fz-juelich.de",
            "@type": "Person",
            "email": "c@fz-juelich.de",
            "name": " Renat Yakupov",
            "givenName": "Renat",
            "familyName": "Yakupov"
          },
          {
            "@id": "https://schema.datalad.org/person#d@fz-juelich.de",
            "@type": "Person",
            "email": "d@fz-juelich.de",
            "name": " Martin Kanowski",
            "givenName": "Martin",
            "familyName": "Kanowski"
          },
          {
            "@id": "https://schema.datalad.org/person#e@fz-juelich.de",
            "@type": "Person",
            "email": "e@fz-juelich.de",
            "name": " Claus Tempelmann",
            "givenName": "Claus",
            "familyName": "Tempelmann"
          },
          {
            "@id": "https://schema.datalad.org/person#f@fz-juelich.de",
            "@type": "Person",
            "email": "f@fz-juelich.de",
            "name": " Stefan Pollmann",
            "givenName": "Stefan",
            "familyName": "Pollmann"
          },
          {
            "@id": "https://schema.datalad.org/person#g@fz-juelich.de",
            "@type": "Person",
            "email": "g@fz-juelich.de",
            "name": " Michael Hanke",
            "givenName": "Michael",
            "familyName": "Hanke",
            "sameAs": "0000-0001-6398-6370"
          }
        ]
      }
    ]
  }
}

$ datalad meta-add -d . metadata_record.json

meta_add(ok): /Users/jsheunis/Documents/psyinf/Data/studyforrest-data (dataset) [added dataset metadata to /Users/jsheunis/Documents/psyinf/Data/studyforrest-data]
```

Author: I used ``meta-conduct`` in a superset to extract metadata from all subdatasets and add it to the git-repo (aka metadata-store) of the superdataset. I think that might be too complex here, since we did not yet describe ``meta-conduct``. For completeness, this was the commands I used:

```shell
> cd superset
> datalad meta-conduct extract_metadata traverser.top_level_dir=$(pwd) traverser.item_type=dataset traverser.traverse_sub_datasets=True extractor.extractor_type=dataset extractor.extractor_name=metalad_studyminimeta adder.aggregate=True
```

If you just want to use the extractor on a specific dataset, you can execute:
```
> datalad meta-extract -d <path-to-dataset> metalad_studyminimeta
``` 
But that would require a ``.studyminimeta.yaml`` file in the dataset. Therefore it might indeed be better to use the core extractor. That would be done with a simple command:

```shell
> datalad meta-extract -d <path-to-dataset> metalad_core
```

or with 

```shell
> datalad meta-extract metalad_core
```
if the current working directory is the root of the dataset from which you want to extract core data.

Reviewer 2: thx, I think I added a very simple example with metalad_core

""""""""""""""""""""""""""""""""""""""""""

Let's start by creating a place where someone else's metadata could live.

.. runrecord:: _examples/DL-101-181-110
   :language: console
   :workdir: beyond_basics/meta

   $ datalad create metadata-assimilation
   $ cd metadata-assimilation

Because MetaLad stores metadata in :term:`Git`'s object store, we use Git to directly fetch metadata from a remote repository, such as this demo on :term:`GitHub`: ``https://github.com/christian-monch/metadata-test.git``.
Because metadata added by MetaLad is not transported automatically but needs to be specifically requested, the command to retrieve it looks unfamiliar to non-Git-users: It identifies the precise location of the :term:`ref` that contains the metadata.

.. find-out-more:: Exactly where is metadata stored?
   :name: fom-metadataobjecttree

   MetaLad employs an internal metadata model that makes the following properties possible:

   * Metadata has a version encoded, but isn't itself version controlled
   * Metadata should not be transported if not explicitly requested
   * It should be possible to only retrieve parts of the overall metadata tree, e.g. certain sub-nodes

>    TODO this needs much more

Reviewer 2: I still need some insights into the metadata organization. @christian-monch do I see correctly that the metadata is organized internally by "dataset_path"? And without -r, dump would only dump one (the first?) dataset_path, or one that is explicitly specified?

Author: That is exactly right!

Here is a longwinded explanation that confirms your view:

Without ``-r`` ``meta-dump`` would only dump metadata for objects specified by the ``DATASET_FILE_PATH_PATTERN``. If ``DATASET_FILE_PATH_PATTERN`` is not provided, an empty string is assumed, which by default translates into a ``tree-file pattern`` (see ``datalad meta-dump --help``) that identifies the topmost superset in the metadata information.

Since there is no studyminimeta-file in the superset, there is also no metadata information and nothing is emitted. But the topmost dataset is still matched by the ``DATASET_FILE_PATH_PATTERN`` and ``-r`` instructs ``meta-dump`` to list metadata for all sub-datasets in the metadata information. So the command line ``datalad meta-dump study-100`` would output metadata because the metadata information contains a dataset at the path "study-100" that has studyminimeta-metadata associated with it.

Reviewer 2: thx!

      "refs/datalad/*:refs/datalad/*"

The metadata is now locally available in the Git repository ``metadata-repo``.
You can verify this by issuing the command ``datalad meta-dump -r``, which will list all metadata in the repository.
Can you guess what type of metadata it contains [#f5]_ ?

.. runrecord:: _examples/DL-101-181-112
   :language: console
   :workdir: beyond_basics/meta/metadata-assimilation

   $ datalad meta-dump -r

Publish metadata to a Git-Repository
""""""""""""""""""""""""""""""""""""

You can also push your metadata to a remote :term:`sibling` (if you have write :term:`permissions`).
This, too, uses a Git command to push only specific :term:`ref`\s.
Let's assume you are in the directory that contains the git repository with your metadata, then you can push your metadata to a remote git repository ``<your repository>``::

    $ git push "<your repository>" "refs/datalad/*:refs/datalad/*"

You will notice that no primary data is stored in the repository ``metadata-destination``. That allows you to publish metadata without publishing the primary data at the same time.

Querying metadata
^^^^^^^^^^^^^^^^^

> TODO: something more about meta-dump and concrete usage example with, e.g., ``jq``

Reviewer 2: Does anyone has an example in their shell history, by any chance? :)

Author: This would create a list of all ``familyName`` values in the ``person`` lists that were extracted from ``.studyminimeta.yaml``-files by ``metalad_studyminimeta``-extractor:

```shell
> datalad meta-dump -r|jq '.extracted_metadata["@graph"][3]["@list"][].familyName'|sort|uniq
```

Reviewer: Something like this? (also studyforrest dataset)

Output array of all subdatasets:

```
$ datalad meta-extract -d . metalad_core > metadata_record_metalad_core.json
$ datalad meta-add -d . metadata_record_metalad_core.json
$ datalad meta-dump . | jq '.extracted_metadata["@graph"] | .[] | select(.["@type"] == "Dataset") | [.hasPart[]? | {"dataset_id": (.identifier | sub("^datalad:"; "")), "dataset_version": (.["@id"] | sub("^datalad:"; "")), "dataset_path": .name}]'

[
  {
    "dataset_id": "d5dd3da0-a631-4c0c-a4a9-de55dfc4620f",
    "dataset_version": "276ceffd8c11db1b99b30b4ae6965aac296fba02",
    "dataset_path": "artifact/3T_movie_eyetracking"
  },
  {
    "dataset_id": "ad9b6c66-4413-4b4f-b6da-b7f25d0d6397",
    "dataset_version": "787741801ad6ff9e4d0aea7a1403bc483f9e1a74",
    "dataset_path": "artifact/3T_structural_mri"
  },
  {
    "dataset_id": "6075c0fa-ab72-4bab-9888-3b597f0e63b1",
    "dataset_version": "64cf42ad00c707bb6533b5944041582c6bfdd384",
    "dataset_path": "artifact/3T_visuallocalizer"
  },
  {
    "dataset_id": "d4759300-5563-467d-be5f-e5b164fb3060",
    "dataset_version": "7b6ec16910eb6a77cd0cc3d2945e7fb292fb9672",
    "dataset_path": "artifact/7T_audiomovie"
  },
  {
    "dataset_id": "c08af312-e05b-43b3-b499-db0d2ad46bf6",
    "dataset_version": "7b58b38af7167d576489cc58e9ec4b3eee0822fc",
    "dataset_path": "artifact/7T_musicperception"
  },
  {
    "dataset_id": "da15d84c-9c8b-11e9-a3fb-f0d5bf7b5561",
    "dataset_version": "f168e8e2b8bef21f373e8a5385e528130ca6339d",
    "dataset_path": "artifact/media"
  },
  {
    "dataset_id": "126cd950-377c-4600-a921-045cf408bd9f",
    "dataset_version": "1b7cb46a97e75aeccab28ef7b61693f4f02a0f8d",
    "dataset_path": "artifact/movie_eyetracking"
  },
  {
    "dataset_id": "0f66b1ba-e9a9-46fd-b9d9-2e64fe94d307",
    "dataset_version": "c74b66cf37c0d4ed8914296c6d7792b2d25696aa",
    "dataset_path": "code/conversion_qa"
  },
  {
    "dataset_id": "7fcd8812-d0fe-11e7-8db2-a0369f7c647e",
    "dataset_version": "2ccaa115543c21e6658950d1cb8cc3038f14272f",
    "dataset_path": "derivative/aggregate_fmri_timeseries"
  },
  {
    "dataset_id": "c8ec2919-493b-4af5-9271-cbe9ebd08c43",
    "dataset_version": "74cd7ec0538448b05fb4d5f91119b279c5e9ab04",
    "dataset_path": "derivative/aligned_mri"
  },
  {
    "dataset_id": "3304e775-5f5f-435a-b68e-d98c9f5fb72a",
    "dataset_version": "aaac44e047d375cd8f791b1b6fe2b739f02c83b2",
    "dataset_path": "derivative/cortical_surfaces_freesurfer"
  },
  {
    "dataset_id": "ceb007ac-ef05-4392-98d2-35c02a774a21",
    "dataset_version": "688d8d8558fe847f4c1b19aed579745bcd6c7744",
    "dataset_path": "derivative/image_space_transformations"
  },
  {
    "dataset_id": "2d05f277-94b0-470b-8e11-4e56691d5b89",
    "dataset_version": "78e04a00fedbe3e055f86c2f9127aa48e1133d55",
    "dataset_path": "derivative/retinotopic_maps"
  },
  {
    "dataset_id": "92e65958-4a5a-4c34-a4f4-ee070f7a123b",
    "dataset_version": "203aa983534fc2b823ff4777a85f4f80d7a68656",
    "dataset_path": "derivative/visual_areas"
  },
  {
    "dataset_id": "5b1081d6-84d7-11e8-b00a-a0369fb55db0",
    "dataset_version": "b623351b43eb2715331ac59ad4cf41682e84ff7d",
    "dataset_path": "original/3T_multiresolution_fmri"
  },
  {
    "dataset_id": "1882e2e6-fbbf-4ade-a65f-3a1615235f51",
    "dataset_version": "e5d2f8368fc5f6717d8ef131041c6d943298d0c7",
    "dataset_path": "original/3T_structural_mri"
  },
  {
    "dataset_id": "3a8648b3-7df8-413f-8efb-4d39040ac174",
    "dataset_version": "10e23aafa8271f742e9022a3522d9a88d7fe30cf",
    "dataset_path": "original/7T_multiresolution_fmri"
  },
  {
    "dataset_id": "5eaff716-54eb-11e8-803d-a0369f7c647e",
    "dataset_version": "72f835ada046bd0479009ea0ff933b30a95b0076",
    "dataset_path": "original/phase2"
  },
  {
    "dataset_id": "4c536c4a-ec61-11e6-9440-00b56d060aa7",
    "dataset_version": "298e2a884d0598dd89753e9b5576bd93d19f335a",
    "dataset_path": "stimulus/computational_annotations"
  },
  {
    "dataset_id": "45b9ab26-07fc-11e8-8c71-f0d5bf7b5561",
    "dataset_version": "29dcce2b9477537b433996ebc342c531139e1d87",
    "dataset_path": "stimulus/curated_annotations"
  }
]
```

Reviewer: Ooh I only saw your reply now @christian-monch. I think yours is a bit more user friendly :)

Author: I think it is good example. Just wanted to mention that the first two lines can be combined to:

```shell
> datalad meta-extract -d . metalad_core  | datalad meta-add -d . --json-lines -
``` 

And since ``-d .`` is equivalent to the default dataset location, the following would be identical:
```shell
> datalad meta-extract metalad_core  | datalad meta-add --json-lines -
```

Reviewer 2: Thanks a lot! I think the family name example would work really well. I get a few errors/warnings by jq in @christian-monch's example - any idea what might be wrong?

```
datalad meta-dump -r | jq '.extracted_metadata["@graph"][3]["@list"][].familyName' | sort | uniq
jq: error (at <stdin>:16): Cannot iterate over null (null)
jq: error (at <stdin>:22): Cannot iterate over null (null)
jq: error (at <stdin>:23): Cannot iterate over null (null)
jq: error (at <stdin>:24): Cannot iterate over null (null)
```

Reviewer 2: This is yet unresolved, I'm merely hiding it in the output. If some has advice, let me know

So to aid with the discoverability of data, one could add metadata to DataLad datasets, extract metadata with MetaLad and multiple extractors, translate extracted metadata to the catalog schema, submit it to ``datalad-catalog`` in order to generate catalog entries, which can all be browsed in a user friendly web-based interface.
Intrigued? Read on to the next section for more information.

Installation
^^^^^^^^^^^^

MetaLad is a stand-alone Python package, and can be installed using

.. code-block:: bash

   pip install datalad-metalad

As with DataLad and other Python packages, you might want to do the installation in a :term:`virtual environment`.

Generally, metadata can either be provided

* manually,
* by running :term:`extractor`\s (``datalad-metalad`` plugins that extract certain metadata from primary data),
* or by any other means that create correct metadata records. For example, you could copy the complete metadata from ``dataset_0`` to ``dataset_1``, by dumping it from one dataset into another::

    $ datalad meta-dump -d dataset_0 -r | \

>       datalad meta-add -d dataset_1 --json-lines -

Reviewer 2: note to self: move to where this belongs
